chain/network/src/routing.rs

```diff
@@ -0,0 +1,375 @@
+        self.peer_forwarding.entry(peer_id).or_insert_with(HashSet::new);
+    }
+
+    /// Add (account id, peer id) to routing table.
+    /// Returns a bool indicating whether this is a new entry or not.
+    /// Note: There is at most on peer id per account id.
+    pub fn add_account(&mut self, account_id: AccountId, peer_id: PeerId) -> bool {
+        self.add_peer(peer_id.clone());
+
+        match self.account_peers.entry(account_id) {
+            Entry::Occupied(_) => false,
+            Entry::Vacant(entry) => {
+                entry.insert(peer_id);
+                true
+            }
+        }
+    }
+
+    pub fn add_connection(&mut self, peer0: PeerId, peer1: PeerId) {
+        self.add_peer(peer0.clone());
+        self.add_peer(peer1.clone());
+        self.raw_graph.add_edge(peer0, peer1);
+        // TODO(MarX): Don't recalculate all the time
+        self.peer_forwarding = self.raw_graph.calculate_distance();
+    }
+
+    pub fn remove_connection(&mut self, peer0: &PeerId, peer1: &PeerId) {
+        self.raw_graph.remove_edge(&peer0, &peer1);
+        // TODO(MarX): Don't recalculate all the time
+        self.peer_forwarding = self.raw_graph.calculate_distance();
+    }
+
+    pub fn register_neighbor(&mut self, peer: PeerId) {
+        self.add_connection(self.raw_graph.source.clone(), peer);
+    }
+
+    pub fn unregister_neighbor(&mut self, peer: &PeerId) {
+        let source = self.raw_graph.source.clone();
+        self.remove_connection(&source, &peer);
+    }
+
+    pub fn sample_peers(&self) -> Vec<PeerId> {
+        // TODO(MarX): Sample instead of reporting all peers
+        self.peer_forwarding.keys().map(|key| key.clone()).collect()
+    }
+}
+
+#[derive(Clone)]
+pub struct Graph {
+    pub source: PeerId,
+    adjacency: HashMap<PeerId, HashSet<PeerId>>,
+}
+
+impl Graph {
+    pub fn new(source: PeerId) -> Self {
+        Self { source, adjacency: HashMap::new() }
+    }
+
+    fn contains_edge(&mut self, peer0: &PeerId, peer1: &PeerId) -> bool {
+        if let Some(adj) = self.adjacency.get(&peer0) {
+            if adj.contains(&peer1) {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    fn add_directed_edge(&mut self, peer0: PeerId, peer1: PeerId) {
+        self.adjacency.entry(peer0).or_insert_with(HashSet::new).insert(peer1);
+    }
+
+    fn remove_directed_edge(&mut self, peer0: &PeerId, peer1: &PeerId) {
+        self.adjacency.get_mut(&peer0).unwrap().remove(&peer1);
+    }
+
+    pub fn add_edge(&mut self, peer0: PeerId, peer1: PeerId) {
+        if !self.contains_edge(&peer0, &peer1) {
+            self.add_directed_edge(peer0.clone(), peer1.clone());
+            self.add_directed_edge(peer1, peer0);
+        }
+    }
+
+    pub fn remove_edge(&mut self, peer0: &PeerId, peer1: &PeerId) {
+        if self.contains_edge(&peer0, &peer1) {
+            self.remove_directed_edge(&peer0, &peer1);
+            self.remove_directed_edge(&peer1, &peer0);
```

Comment:
Either check `contains_edge` for both, or for none?
It's weird to check the existence of the edge in one direction but to remove from both.

Reply:
But it is an undirected graph